### ISSUE TITLE
Overhaul of the stats module

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#20643]
The _stats_ module is about 10 years old and has been expanded a lot in the last 2 years, with contributions from many authors.

The code complexity keeps getting more and more tangled, so it's time to review the class structure and design of the _stats_ module in order to clean up the code, before it gets even more intricated.

This issue is meant to collect possible proposals in order to clean up the code (just cleaning up and keeping the existing functionality, no new features here).

Some examples:

1. [ ] Type instability in subclasses of `PSpace`: `FinitePSpace` accepts `(domain, set)` as parameters, while `ContinuousPSpace` accepts `(symbol, set)` as parameters. This is a design flaw dating back 10 years, which has never been corrected.
2. [ ] Computation of integrals/summations is sometimes implemented in subclasses of `RandomDomain`, sometimes in subclasses of `PSpace`. Is that a consistent approach?
3. [ ] Indexed random variables are sometimes represented with `Indexed`, sometimes with `RandomIndexedSymbol`. Should we rather use only one class?
4. [ ] `RandomIndexedSymbol` subclasses `RandomSymbol` but changes the types of the parameters. Usually in SymPy when you define a subclass, you should keep the same parameters of the superclass in the constructor (only difference allowed in adding new parameters at the end of the parameter list).
5. [ ] Do we need the class structure distringuishing single/multivariate and finite/discrete/continuous to be replicated as subclasses of `RandomDomain`, `PSpace` and `Distribution`? Wouldn't it be simpler to just subclass `Distribution` and implement all the logic there? This also complicates the way multivariate RVs work. Maybe we should rather use generic `RandomDomain` classes to keep track of finite/discrete/continuous RVs and the integration/summation limits?
6. [ ] Finite and discrete distributions implement `.pdf`. It would make more sense to call it `.pmf`, as it is more consistent with the usual terminology. `.pdf` is OK for finite/discrete distributions if we use `DiracDelta`.
7. [ ] Mark every returned type and enforce static typecheck with `mypy`. This will make the code more consistent and type-stable. Another way is to add `assert isinstance(...)` in the code, to make sure the types are consistent at runtime. Python allows to much freedom when writing the code and may end up making everything messy.

**Comment by Upabjojr** [comment]
@czgdp1807 feel free to fill this list.

**Comment by Upabjojr** [comment]
Concerning the dimensionality of the probability spaces, we have:

1. single univariate random variable: one-dimensional probability space.
2. multivariate random variable: N-dimensional probability space. These can be regarded as:
  a. products of independent random variables,
  b. joint random variables.
3. discrete-time stochastic processes: countably infinite dimensional probability spaces.
4. continuous-time stochastic processes: uncountably infinite dimensional probability spaces.

Both points 3. and 4. have complicated relationships regarding dependence/independence, stemming from the type of stochastic process they represent.

**Comment by Upabjojr** [comment]
Regarding the simplification of random expressions, `PSpace` should be able to _remember_ the origin of the RV.

For example, suppose `N` is normal, we know that `exp(N)` is lognormally-distributed. If we are going to simplify `exp(N) ==> LN` where `LN` is a lognormally-distributed `RandomSymbol`, we have to find a way to represent in its `PSpace` the fact that both `N` and `LN` are operating on the same probability space.

How should `PSpace` be modified in order to handle such a case?

**Comment by oscarbenjamin** [comment]
> Regarding the simplification of random expressions, `PSpace` should be able to _remember_ the origin of the RV.

If simplification was applied to distributions rather than expressions or pspaces then this issue is sidestepped in a clean way:
```
>>> X = Normal('X', 0, 1)
>>> distribution(exp(X))
TransformedDistribution(Lambda(x, exp(x)), NormalDistribution(0, 1))
>>> simplify(distribution(exp(X))
LogNormal(0, 1)
```
Remembering which random symbol is which can be done by the expression. On converting the expression to a distribution the random symbols all become dummy symbols and it isn't necessary to remember any relationship to anything outside of the distribution object.

**Comment by Upabjojr** [comment]
> Remembering which random symbol is which can be done by the expression.

I had another idea... what about redefining `RandomSymbol` as:

```python
RandomSymbol(name: str, probability_space: Dummy, random_domain: RandomDomain, distribution: Distribution)
```
I mean... abolish the class `PSpace` and make it a dummy. All of its methods should then be moved somewhere else. This is more consistent with the [mathematical definition of random variable](https://en.wikipedia.org/wiki/Random_variable#Definition). The _probability space_ is just a hidden space upon which a probability measure can be defined. What we call _random variables_ are just measurable functions mapping to the real numbers. The mapping is not part of the `PSpace` as per the definition of RVs. The `PSpace` could just be a dummy object to reason about dependencies.

Another possibility would be to introduce a `RandomMapping` object, binding the _probability space_, _random domain_ and _distribution_.

Unfortunately these approaches require a lot of code rewriting in the _stats_ module, so they should be thought of carefully.

Concerning distributions, remember that Wolfram Mathematica does not support subclassing, so you cannot define a `RandomSymbol` object there. You need to use ordinary symbols and bind them to distributions. SymPy's approach is completely different, distributions are hidden internal objects... end-users should use expressions of `RandomSymbol` objects.

**Comment by oscarbenjamin** [comment]
> I had another idea... what about redefining `RandomSymbol` as:

We still need a distribution for the expression as a whole `X + Y` will be an `Add` and we can't attach probability spaces and distributions to the `Add`.

**Comment by Upabjojr** [comment]
> We still need a distribution for the expression as a whole `X + Y` will be an `Add` and we can't attach probability spaces and distributions to the `Add`.

If `X` and `Y` have different probability spaces, then `pspace(X + Y)` is a product probability space (as it is already right now). If `X` and `Y` have the same underlying probability space, then `pspace(X + Y)` should return a single probability space. Unfortunately this cannot be done right now, as we have no way of representing more distributions with the same underlying probability space.

**Comment by czgdp1807** [comment]
It's great that this issue has been opened. I agree with @Upabjojr at many places in the list in the OP. I think that the design of `stats` is slightly inconsistent. 

1. As of now, if any public API of stats module is called, it goes through a number of steps. For example, in the case of `density`, first, `pspace` is called on the input expression and then the task of computing density is delegated to `compute_density` of the pspace returned by `pspace` function. Now, that `compute_density` method gives the results by somehow calling, `density` property which is in-turn is provided at the time of initializing the pspace. And some sub-classes of `PSpace` like `SinglePSpace` have a `distribution` property as well.
2. The Wikipedia definition of a random variable provides some hints on how we can characterize the class, `RandomSymbol`. Each RV has a possible set of outcomes termed as sample space and if I am not wrong it same as `RandomDomain`, an event space, and a distribution function defining what is the probability of each possible outcome. The set of possible outcomes may not necessarily be real or complex, it can be anything. 
3. Going by the above analysis I would suggest to re-write the `PSpace` class. It should contain, the sample space, reference to the distribution as it's attributes and `compute_density`, `compute_mgf` and other such methods implementing generic computation algorithms. It should be possible to define the pspace on any random expression. Now, the `Distribution` class should have a generic function called, `distribution_function` instead of `pdf`/`pmf` and a `cumulative_distribution_function`(in short cdf), both would return `None`. This suggests that we don't need `RandomDomain` or any class to represent `domain` as it would be already stored by `PSpace` class. The attributes of `Distribution` class can contain the support of `distribution_function` over which it is defined.
4. Also, by the above design, classes like, `NormalDistribution`, `BinomialDistribution` can directly inherit `Distribution` class to define their own properties rather than inheriting `SingleContinuousDistribution`/`ContinuousDistribution`. We can also define classes like `NormalPSpace`, `BernoulliPSpace`, etc. to define pspaces over normal RV, bernoulli RV etc and they can define their own specific algorithms inside `compute_density`, `compute_mgf`, etc.
5. Now, after the above design pipeline would be `density(say Normal RV)` -> `pspace` (returns `NormalPSpace`) -> calls `NormalPSpace.compute_density` -> fetches pdf from `NormalDistribution.distribution_function` and applies `Piecewise` over domain -> final output.
6. Some specific pspace classes may be kept like, a pspace class for multivariate expressions but consistency should be followed i.e., they should override similar methods.
7. `RandomSymbol` should only contain reference to `PSpace` object as an attribute and nothing else and rest should be managed by that `PSpace`. We can also define `RandomExpr` and associate a `PSpace` class for it. This should be done by `pspace` function. 

The above points just intend to simplify the design and remove redundant classes. I may be wrong or would have assumed several things. Please feel free to point that out.

**Comment by Upabjojr** [comment]
Maybe _distribution_ and _density_ should be unified as the same concept. There is a code design problem though, _distribution_ classes require the density (i.e. `pdf`) to be defined as a method, while _density_ objects accept density as a parameter.

I would also like to get rid of the distinction between CRV, FRV and DRV subclasses. In my opinion, those distinctions should all be handled by `RandomDomain`.

**Comment by czgdp1807** [comment]
Is `RandomDomain` same as sample space? Just confirming.
If we want to experiment with redesigning `stats` module, may be we can create a module like `stats_new` and experiment with designs. In addition, whenever users import something from `stats` module, we can make them aware about in a `stats_new` in a message(may be a `UserWarning`) and request them to try out the APIs from `stats_new` and give us feedback. This way, no drastic changes will have to be made in original `stats` module and we will be replace it with `stats_new` whenever it gets stabilized.

**Comment by Upabjojr** [comment]
> If we want to experiment with redesigning `stats` module, may be we can create a module like `stats_new` and experiment with designs

Well, forking is usually a bad strategy. There is a guy who forked SymPy a few years ago and kept working on it. I tried to see if we can import back his improvement to the Gruntz algorithm, but the code has changed too much and the task is quite hard.



> This way, no drastic changes will have to be made in original `stats` module and we will be replace it with `stats_new` whenever it gets stabilized.

Drastic changes are OK as long as they are in the private API. The criterion is simple, everything that is imported with `from sympy.stats import *` should not change its behaviour, the rest can.

**Comment by czgdp1807** [comment]
> The criterion is simple, everything that is imported with `from sympy.stats import *` should not change its behaviour, the rest can.

I see, then that's great. I will try to introduce some simplifications in the current design by maintaining the above constraint. Will submit a draft PR for this.

**Comment by rikardn** [comment]
As a user of the `sympy.stats` module I like the idea of this issue. I just want to add some issues I have been struggling with:

1. Currently joint distributions must be indexed. I would like to be able to say something like: `X1, X2 = Normal(['X_1', 'X_2'], [0, 0], [[1, 0], [0, 1]])` where each random variable can have its own symbol.
2. Both multivariate normal and single variable normal are created using the same function `Normal`, but the names of their distribution parameters are different `mu` and `sigma` for the multivariate and `mean` and `std` for the single. This is a bit confusing.
3. Sometimes you would like to use different parametrizations of a distribution and even convert between different parametrizations. For example (mu, sigma^2) or (mu, sigma) or even (mu, tau) for the normal distribution. Some way of allowing different parametrizations and/or reparametrizing distributions would be useful.
4. Allowing symengine matrices as input to the joint normal distribution. I have been thinking of submitting a PR for this. A symengine matrix is blocked by the init code.
5. Substituting values for random variables cannot be done with a basic dict: `expr.subs({'X': 23})`. I know that the reason is that X will become a regular Symbol after sympifying, but it is inconvenient for working with expression that are a mix of rvs and regular symbols to keep track of which is which when substituting. Ideally subs shouldn't differentiate between the types of the symbols only the name.

And finally: RandomVariables are super! It is very intuitive to mix them into an expression and sample, calculate the expected value etc.

**Comment by czgdp1807** [comment]
I agree with point 2. It highlights an ambiguity which can be removed and should be removed though will affect users with the API change. `3` is a nice feature request if it isn't already possible using existing SymPy APIs. I am not very sure about other points.

**Comment by Maelstrom6** [comment]
Is there a fork or a Google Drive link that we could use to lay out the structure of the new stats module? Like, it should have a list of all the new classes, their arguments and their properties. Something like:

```
rv.py
├───RandomSymbol
│   ├───__new__(cls, name: str, probability_space: Dummy, random_domain: RandomDomain, distribution: Distribution) -> RandomSymbol
│   ├───free_symbols(self) -> set
│   └───...
...
```

That is, we don't need to write any code just yet. All we need is what the structure looks like and what the goal of each function is.

**Comment by Maelstrom6** [comment]
About half of the issues raised about the stats module is actually an issue with the integrate module. 

Should we try to improve the integrate module to handle piecewise expressions and DiracDelta expressions better? Or should we ease up on the integrate module and let the stats module take advantage of the computationally simpler theorems related the field of statistics?

I suggest we make the integrate module do the heavy lifting while we set up the basics of the new stats module. But I just want to confirm if this is the consensus before we start.

**Comment by oscarbenjamin** [comment]
We should definitely try to improve the `integrate` module. On the other hand the stats module does sometimes pass unnecessarily complicated expressions to integrate and that should *also* be fixed.

Something that would be helpful would be to make a new issue that captures all of the examples where stats fails because of failures in integrate. The different examples could be listed in a standard way that explains what the problems are. As new issues are opened the examples can be moved there. That would make it easier to see all the examples in one place. Then we can close all of the other stats issues and just have one that clearly explains what the problems are.

**Comment by Maelstrom6** [comment]
I have a comment somewhere with an almost-complete list. I'll see if I can set up the issue tomorrow.

**Comment by asmeurer** [comment]
> Finite and discrete distributions implement .pdf. It would make more sense to call it .pmf, as it is more consistent with the usual terminology. .pdf is OK for finite/discrete distributions if we use DiracDelta.

I don't do enough probability, so maybe others can comment here. Is this sort of thing confusing to people? While it may be technically less correct to use `pdf` for a discrete distribution, having a uniform API is more important, especially if the only real reason for splitting the API is a difference in underlying names. Is the suggestion just to add `pmf` as an alias for `pdf`?

+1 on improving `integrate` whenever possible. If we improve `integrate`, then it will be able to compute the corresponding integrals even if they don't come directly from the stats module, for instance, if someone just writes down a probability integral by hand. Integrals that look like expectations can also come up in contexts outside of probability, so improving `integrate` will be beneficial to a lot more people. Integrals with deltas in particular come up in all kinds of contexts outside of probability, and improving them should be a priority even if we didn't have sympy.stats. For example, a "table lookup
 in stats could just as well be a table lookup in something like manualintegrate, and in fact they can often even be generalized beyond the integrals stats would produce. 

However, it's not always so easy to improve `integrate` generically, so it may be necessary to work around the current limitations. But whenever that is done, it should be treated as a bug `integrate`.

**Comment by oscarbenjamin** [comment]
> I don't do enough probability, so maybe others can comment here. Is this sort of thing confusing to people? While it may be technically less correct to use `pdf` for a discrete distribution, having a uniform API is more important, especially if the only real reason for splitting the API is a difference in underlying names. Is the suggestion just to add `pmf` as an alias for `pdf`?

In some formulations of stats discrete and continuous distributions are treated separately with one being based on PMF and summations and the other on PDF and integrals. It is also possible to define a framework that incorporates both where a discrete probability distribution is treated as a continuous distribution with a discontinuous CDF and a PDF that is made of delta functions.

The issue is that there can be a meaning for the PDF of a discrete distribution that is very different from the PMF. Combining the PDF of a discrete distribution and the PMF of a continuous distribution in a uniform API would be useful if they were interchangeable in any situation but they are not: I can't integrate a PMF and I can't sum a PDF so in what situation do I not care which I have?

**Comment by Maelstrom6** [comment]
I've started to add some bare bones to a new stats module in [my branch](https://github.com/Maelstrom6/sympy/tree/stats_overhaul). Just an explanation of classes and methods and some type hints.

But I'm not sure what to do with the `PSpace` and `RandomDomain` discussion. I think we need to remove one of them, so that we don't need to create two new classes every time we need to define a new kind of random variable (like single or continuous or product). Which one should we remove?

**Comment by czgdp1807** [comment]
I am a bit busy these days but still commenting on your latest query as it seems like a blocker. I believe RandomDomain should be removed as it's redundant. A simple domain attribute addition in PSpace class would make life easier. PSpace is the class which implements most of the generic and distribution type specific computational algorithms, so it is important and removing it may create troubles moving those implementations to other places.

**Comment by Upabjojr** [comment]
> But I'm not sure what to do with the `PSpace` and `RandomDomain` discussion. I think we need to remove one of them, so that we don't need to create two new classes every time we need to define a new kind of random variable (like single or continuous or product). Which one should we remove?

That's the problem. Designing a better API is not as simple as it looks like... PSpace should be made compatible with univariate random variables, multivariate, multivariate independent RVs, multivariate joint RVs, multivariate compound RVs, discrete time stochastic process RVs and continuous time stachastic process RVs. Ideally the equality / inequality reducer for the domain should be as general as possible.

I'm also quite busy right now, and I'm currently concentrating on another SymPy module. I hope I will be able to fix this eventually.

**Comment by thambi999** [comment]
U All are busted, spoil ppl life, business, future, more then that u all how u can make anything for money mother fucker, On 10 Feb 2021 03:41, Francesco Bonazzi <notifications@github.com> wrote:

But I'm not sure what to do with the PSpace and RandomDomain discussion. I think we need to remove one of them, so that we don't need to create two new classes every time we need to define a new kind of random variable (like single or continuous or product). Which one should we remove?

That's the problem. Designing a better API is not as simple as it looks like... PSpace should be made compatible with univariate random variables, multivariate, multivariate independent RVs, multivariate joint RVs, multivariate compound RVs, discrete time stochastic process RVs and continuous time stachastic process RVs. Ideally the equality / inequality reducer for the domain should be as general as possible.
I'm also quite busy right now, and I'm currently concentrating on another SymPy module. I hope I will be able to fix this eventually.
